auditing/order_fulfillment.py

```diff
@@ -0,0 +1,162 @@
+    JOIN order_line ol ON (ol.order_id = o.id)
+    JOIN catalogue_productattributevalue pav ON (pav.product_id = ol.product_id)
+    JOIN catalogue_productattribute pa ON (pa.id = pav.attribute_id AND pa.code = 'course_key')
+    JOIN ( SELECT pav2.product_id, pav2.value_text
+    FROM catalogue_productattributevalue pav2
+    JOIN catalogue_productattribute pa2 ON (pa2.id = pav2.attribute_id AND pa2.code = 'certificate_type')
+    ) AS modes ON (modes.product_id = ol.product_id)
+    LEFT JOIN refund_refundline rl ON (rl.order_line_id = ol.id)
+WHERE
+    o.status = 'Complete'
+    AND o.date_placed > %s
+    AND rl.id IS NULL
+ORDER BY
+    u.username ASC
+    , pav.value_text ASC
+    , o.date_placed ASC;
+"""
+    date_placed = date_placed.strftime('%Y-%m-%d %H:%M:%S')
+    orders = pandas.read_sql_query(sql, ecommerce_db, index_col='number', params=(date_placed,),
+                                   parse_dates=('date_placed',))
+    orders.reset_index(level=0, inplace=True)
+    original_order_count = len(orders)
+
+    # Identify upgrades and remove the honor purchase.
+    # TODO We will need to do the same for credit.
+    grouped = orders.groupby(['username', 'course_id'])
+    for key, row_ids in grouped.groups.iteritems():
+        username, course_id = key
+        if len(row_ids) > 1:
+            row_ids.sort()
+            honor_order = orders.ix[row_ids[0]]
+            verified_order = orders.ix[row_ids[1]]
+
+            logger.info(
+                'User [%s] upgraded for course [%s]. Honor order [%s] will be ignored in favor of verified order [%s].',
+                username, course_id, honor_order['number'], verified_order['number'])
+            orders = orders.drop(row_ids[0])
+
+    logger.info('Dropped [%d] orders.', original_order_count - len(orders))
+    return orders
+
+
+def get_enrollments(usernames, course_ids):
+    """ Retrieve all enrollments for the given courses and users. """
+    sql = """SELECT
+    u.username
+    , e.course_id
+    , e.mode
+    , e.is_active
+FROM
+    student_courseenrollment e
+    JOIN auth_user u ON (u.id = e.user_id)
+WHERE
+    u.username IN ({usernames})
+    AND e.course_id in ({course_ids});
+""".format(usernames=','.join('%s' for __ in usernames), course_ids=','.join('%s' for __ in course_ids))
+
+    params = list(usernames) + list(course_ids)
+    enrollments = pandas.read_sql_query(sql, edxapp_db, params=params)
+    return enrollments
+
+
+def identify_missing_enrollments(orders, enrollments):
+    """ Identify the orders that do not have corresponding enrollments. """
+    merged = pandas.merge(orders, enrollments, on=['username', 'course_id', 'mode'], how='left')
+    unfulfilled_orders = merged[merged.is_active.isnull()]
+    return unfulfilled_orders
+
+
+def run_audit():
+    date_placed = datetime.utcnow() - timedelta(minutes=ORDER_WINDOW_START_TIME)
+    orders = get_orders(date_placed)
+    usernames = pandas.unique(orders.username.ravel())
```

Comment:
I typically just do: `orders.username.unique()`

Is that consequentially different from using `ravel()`?